src/main.cpp

```diff
@@ -90,87 +91,206 @@ namespace
-                    // The "admin_only" flag supersedes the "editors" configuration option.
-                    if (config->count("admin_only") && config->at("admin_only").get<bool>()) {
-                        return user_is_administrator(*rei.rsComm);
-                    }
-
-                    namespace adm = irods::experimental::administration;
-
-                    const adm::user user{rei.uoic->userName, rei.uoic->rodsZone};
-
-                    for (auto&& editor : config->at("editors")) {
-                        if (const auto& type = editor.at("type").get_ref<const std::string&>(); type == "group") {
-                            const adm::group group{editor.at("name").get_ref<const std::string&>()};
-
-                            if (adm::server::user_is_member_of_group(*rei.rsComm, group, user)) {
-                                return CODE(RULE_ENGINE_CONTINUE);
-                            }
-                        }
-                        else if (type == "user") {
-                            if (editor.at("name").get_ref<const std::string&>() == adm::server::local_unique_name(*rei.rsComm, user)) {
-                                return CODE(RULE_ENGINE_CONTINUE);
-                            }
-                        }
-                    }
-
-                    // At this point, the user is not an administrator and they aren't a member of
-                    // the editors list. Therefore, we return an error because the user is attempting to
-                    // modify metadata in a guarded namespace.
-                    return ERROR(CAT_INSUFFICIENT_PRIVILEGE_LEVEL, "User is not allowed to modify metadata");
-                }
-            }
-        }
-        catch (const json::exception&) {
-            // clang-format off
+		_exists = (_rule_name == "pep_api_mod_avu_metadata_pre") ||
+		          (_rule_name == "pep_api_atomic_apply_metadata_operations_pre");
+		return SUCCESS();
+	}
+
+	auto list_rules(irods::default_re_ctx&, std::vector<std::string>& _rules) -> irods::error
+	{
+		_rules.push_back("pep_api_mod_avu_metadata_pre");
+		_rules.push_back("pep_api_atomic_apply_metadata_operations_pre");
+		return SUCCESS();
+	}
+
+	auto check_operation_for_violations(const std::string& _object_name, ruleExecInfo_t& rei, const json& config) -> irods::error
+	{
+		const auto prefixes_iter = config.find("prefixes");
+		if (prefixes_iter == config.cend()) {
+			return CODE(RULE_ENGINE_CONTINUE);
+		}
+		const auto prefix_matched =
+			std::any_of(prefixes_iter->cbegin(), prefixes_iter->cend(), [&_object_name](const json& prefix) {
+				return boost::starts_with(_object_name, prefix.get_ref<const std::string&>());
+			});
+		if (!prefix_matched) {
+			return CODE(RULE_ENGINE_CONTINUE);
+		}
+		const auto admin_iter = config.find("admin_only");
+		if (admin_iter != config.cend() && admin_iter->get<bool>()) {
+			return user_is_administrator(*rei.rsComm);
+		}
```

Comment:
I thought these lines are dependent on clearing the guard clause checking prefix_matched (L116-), since it shouldn't happen unless the operation is trying to write metadata to a protected namespace.

Reply:
Yes, your logic is correct, but I think we can improve this a little more. Consider the following ...
- When `admin_only` is true, we return immediately
    - Here, we only pay for the prefix check
    - Great, we didn't waste that many extra cycles
- When `admin_only` is not defined or is false, the plugin processes the `editors` list
    - Here, we pay for the prefix check and the editors check
    - **These can potentially run for every operation**

If you think about it, the `prefixes` list doesn't matter at all when `admin_only` is true.

By checking the `admin_only` option earlier, we don't pay for the prefix check, we don't pay for the admin flag check for every operation, and the behavior I explained becomes more obvious.

This can be applied to the other PEP too, assuming my understanding is correct.

Thoughts? Did I miss anything?

Reply:
i agree.  get out early if admin_only is true.